File: Tokens.swift
/* TOKENS */

let letters = CharacterSet.letters

enum TT {
    case FACTOR 
    case OPERATOR
    case GROUP
    
    case STRING

    case IDENTIFIER
    case EOF
    case INDICATOR
    
    case EQ
    case EE 
    case NE 
    case LT 
    case GT 
    case LOE 
    case GOE 

    case NOT
    case OR 
    case AND 

    case IF
    case ELIF 
    case ELSE 
    case FOR
    case IN
    case WHILE

    case LCURLY 
    case RCURLY 

    case FUNC
    case ARROW
    case COMMA
}

let TT_INT       = "INT"
let TT_FLOAT     = "FLOAT"
let TT_STRING    = "STRING"

let TT_PLUS      = "PLUS"
let TT_MINUS     = "MINUS"
let TT_MUL       = "MUL"
let TT_DIV       = "DIV"
let TT_POW       = "POW"

let TT_LPAREN    = "LPAREN"
let TT_RPAREN    = "RPAREN"
let TT_LCURLY = "LCURLY"
let TT_RCURLY = "RCURLY"

let TT_ID        = "IDENTIFIER" // name of variables
let TT_EOF       = "EOF"
let TT_INDICATOR = "INDICATOR"

let TT_EQ        = "EQ"
let TT_EE        = "EQUALS"
let TT_NE        = "NOT EQUALS"
let TT_LT        = "LESS THAN"
let TT_GT        = "GREATER THAN"
let TT_LOE       = "LESS THAN OR EQUALS"
let TT_GOE       = "GREATER THAN OR EQUALS"

let TT_NOT       = "NOT"
let TT_AND       = "AND"
let TT_OR        = "OR"

let TT_IF        = "IF" 
let TT_ELIF      = "ELSE IF"
let TT_ELSE      = "ELSE"
let TT_FOR       = "FOR"
let TT_WHILE     = "WHILE"
let TT_IN        = "IN"

let TT_FUNC      = "FUNC"
let TT_COMMA     = "COMMA"
let TT_ARROW     = "ARROW"



class Token {
    // This is is Metatype, (ex: factor, operator, etc)
    var type: TT 
    // This is the name of the type (ex: int, add, minus, etc)
    var type_name: String 
    var value: Any?
    var pos: Position?

    init() {
        self.type = .FACTOR 
        self.type_name = ""
        self.value = ""
        self.pos = nil
    }

    init(type: TT, type_name: String, value: Any?=nil, pos: Position?=nil) {
        self.type = type
        self.type_name = type_name
        self.value = value
        self.pos = pos
    }

    func as_string() -> String {
        return ("\(self.type) : \(self.value ?? "")")
    }
}

File: tokens.py
TT_INT = "INT"
TT_STRING = "STRING"

MT_FACTOR = "FACTOR"
MT_NONFAC = "NON FACTOR"

TT_PLUS = "PLUS"
TT_MINUS = "MINUS"
TT_MUL = "MUL"
TT_DIV = "DIV"
TT_POW = "POW"

TT_LCURLY = "LCURLY"
TT_RCURLY = "RCURLY"
TT_LPAREN = "LPAREN"
TT_RPAREN = "RPAREN"

TT_ID = "IDENTIFIER"
TT_EOF = "EOF"
TT_COLON = "COLON"

TT_EQ = "EQ" # This the assignment equals 
TT_EE = "EQUALS" # This is the comparison equals 
TT_NE = "NOT EQUALS"
TT_LT = "LESS THAN"
TT_GT = "GREATER THAN"
TT_LOE = "LESS THAN OR EQUAL"
TT_GOE = "GREATER THAN OR EQUALS"

TT_NOT = "NOT"
TT_AND = "AND"
TT_OR = "OR"

TT_IF = "IF"
TT_ELIF = "ELIF"
TT_ELSE = "ELSE"
TT_FOR = "FOR"
TT_WHILE = "WHILE"
TT_IN = "IN"

TT_FUNC = "FUNC"
TT_COMMA = "COMMA"
TT_ARROW = "ARROW"

class Token:
    def __init__(self, metatype="", type="", value="", pos=None):
        self.metatype = metatype
        self.type = type
        self.value = value 
        self.pos = pos 
    
    def as_string(self):
        return (f'{self.type} : {self.value}')

class Position:
    def __init__(self, ln=0, col=0, fn="repl"):
        self.ln = ln 
        self.col = col 
        self.fn = fn